### PR TITLE
[UNR-5121] Cross server possession tests can't be placed in the same map

### PR DESCRIPTION
**Pull request by razvan-at-improbable**
#### Description
Added cleanup steps for the tests so that the map is reverted to its original state at the end of each test. In all cases this involved restoring possession of the original pawns to the player controllers. For the lock test needed to add support to release locks from netdriver.
Some refactoring was done as there's shared code between the tests and they share a parent class. 
Some fixes were applied to the tests themselves as some didn't achieve their intended functionality.

#### Tests
How did you test these changes prior to submitting this pull request?
Ran the tests in CI and on multiple machines multiple times through the Session Frontend.
What automated tests are included in this PR?
All Cross-Server Possession Tests.
STRONGLY SUGGESTED: How can this be verified by QA?
Starting the tests through the Session Frontend by going to Window > Generate test maps, then through Session Frontend select all the tests in CrossServerPossessionMap and start them with 1 run multiple times. (This might flake due to "A functional test is already running" error if tried with multiple runs, which is a known bug found at https://improbableio.atlassian.net/browse/UNR-5171)
#### Primary reviewers
@improbable-valentyn 
I suggest going through each commit separately

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerMultiPossessionTest.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerMultiPossessionTest.cpp:70]: As per previous PR. Why do we need the DeltaTime lambda here? (Instead of the other parameter which doesnt take a float)

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionLockTest.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionLockTest.cpp:69]: Looks like there's only locks on the controllers, no?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionLockTest.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionLockTest.cpp:73]: Shouldn't only the server with authority do this?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:55]:
```suggestion
					if (!Pawn->HasAuthority())
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:53]: This assert seems redundant (it will always pass because `PlayerController->HasAuthority()` is checked in the `if` above)

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:67]: Could you add an explicit comment that the pawn is expected to be spawned on Server 4 (or rather located in its area)?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:55]: Nice catch this

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:67]: Why this server change?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:68]: Why is this on tick?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:85]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:53]: Yeah weird, I would suggest removing PlayerController->HasAuthority() check in the if, but its difficult to guess if this was the original intention.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:75]: This feels a bit weird, this is testing the assumptions the test itself makes rather than the tested scenario if that makes sense? If we want to be defensive here we should use a normal unreal assert: `check()`

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionPlayerController.h
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionPlayerController.h:17]: Doesn't look like this is used

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (simonsarginson) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:68]: +1

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:53]: I think the one in the if is right, because we want this logic to only execute on the server with auth over the controller.

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:57]: temp logging

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:91]: temp logging

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:18]: PlayerController (consistent with elsewhere?)

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.h:47]: PlayerController

**Review by improbable-valentyn — APPROVED** [review]
Looks good! (please address the suggestions before merging)

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionPlayerController.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/TestPossessionPlayerController.cpp:85]: `NetDriver != nullptr` is this ever legit? I don't think we can run this if we are not running on spatial (which is the only valid reason for this to be null I can think of in this context). Could be a `check`

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (razvan-at-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:53]: Yup the assert is indeed redundant. It was in the original test which I wasn't comfortable changing when I was working on this file 😄

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (razvan-at-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:67]: The initial test was passing because the assert was never reached because the condition in the "if" was never true, as the PlayerController migrates to Server 4 (that's where the pawn is spawned) so it would be nullptr when accessed from Server 1.

Comment (razvan-at-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:67]: Yup

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (razvan-at-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:75]: This was part of our temp debugging and I forgot to remove it 😅

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (razvan-at-improbable) [reply SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:68]: Everything was on tick by default, working to change that now

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp
Comment (simonsarginson) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/CrossServerPossessionTest.cpp:43]: This shouldnt have DeltaTime param

**Review by simonsarginson — APPROVED** [review]
LGTM with some comments

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp
Comment (mironec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestPossession/SpatialTestRemotePossession.cpp:54]: Maybe we could be slightly more strict here, e.g. asserting that the pawn is not null and that we have authority over it IF we have authority over the playerController.

**Review by mironec — COMMENTED** [review]
Left a few comments, and I generally approve.
But needs an approving review from someone other than me, and also would encourage to run this a few more times, to see if anything is maybe broken.

**Pull request merged**
